// __BEGIN_LICENSE__
// Copyright (C) 2006, 2007 United States Government as represented by
// the Administrator of the National Aeronautics and Space Administration.
// All Rights Reserved.
// __END_LICENSE__

package vw.platefile;

// ----------------------------------------------------------------------
//                     Blob file Data Structures
// ----------------------------------------------------------------------

// Each entry in the blob file is described by a BlobRecord.  This
// allows us to save and access data and headers with variable
// positioning and length.  BlobRecords are stored in the blob files:
// <platefile>/plate_<#>.blob
message BlobRecord {
  required uint32 header_offset = 1;
  required uint32 header_size = 2;
  required uint32 data_offset = 3;
  required uint32 data_size = 4;
}

// The TileRecord store metadata about the tile itself as well as the
// information needed to place the tile back into the index tree.  The
// [col, row, depth, transaction_id] info can be used to reconstruct the index
// by reading TileRecords in the blob files.
message TileHeader {
  required int32 col = 5;
  required int32 row = 6;
  required int32 depth = 7;
  optional uint64 transaction_id = 8 [ default = 0 ];
  optional string filetype = 10 [ default = 'jpg' ];
}

// ----------------------------------------------------------------------
//                      Index Data Structures
// ----------------------------------------------------------------------

enum IndexRecordStatusType {
  INDEX_RECORD_EMPTY = 1;
  INDEX_RECORD_STALE = 2;
  INDEX_RECORD_VALID = 3;
  INDEX_RECORD_PRIMED = 4;
}

// The IndexHeader store basic metadata for the platefile, such as the
// default tile size, file type, and map projection information.
// This is typically stored in the <platfile>/plate.index file.
message IndexHeader {

  // "Public" Metadata
  //
  // You should set these values before passing an IndexHeader along
  // to the Index::construct_create() method.
  //
  optional uint32 tile_size = 2;     // tile width/heigh in pixels 
  optional string tile_filetype = 3; // file type [e.g. tif, png, or jpg]
  optional int32 pixel_format = 4;   // see vw/Image/PixelTypeInfo.h
  optional int32 channel_type = 5;   // see vw/Image/PixelTypeInfo.h
  optional string type = 6;          // platefile type [ toast, kml, or gigapan ]
  optional string description = 7;   // textual description

  // Privately Managed Metadata
  //
  // These values are managed privately be the actual index class. You
  // should never need to set these yourself.
  //
  required int32 platefile_id = 1;   // a unique number that identifies this platefile
  optional int32 version = 10 [default = 1];  
  optional int32 transaction_read_cursor = 11 [default = 0];
  optional int32 transaction_write_cursor = 12 [default = 1];
  repeated int32 complete_transaction_ids = 13 [packed = true];

}

// The IndexRecord stores basic metadata for locating a tile in a
// blob.  The IndexRecord is used to store this information in a live
// index (in main memory), and can also be stored as part of a
// TileHeader, which is written to the blob file on disk.  Other
// metadata about that tile is stored alongside the tile itself in the
// blobfile in a TileRecord protobufer (see below).
message IndexRecord {
  optional int32 blob_id = 1 [default = -1];
  optional uint64 blob_offset = 2 [default = 0];
  optional IndexRecordStatusType status = 3 [ default = INDEX_RECORD_EMPTY ];
}


// ----------------------------------------------------------------------
//             Remote Procedure Calls (RPC) for Index Server
// ----------------------------------------------------------------------

// ----------------------------------
// Client --> Server Requests
// ----------------------------------

message IndexOpenRequest {
  required string plate_name = 1;
}

message IndexCreateRequest {
  required string plate_name = 1;
  required IndexHeader index_header = 2;
}

// If given, the arguments act as filters
message IndexListRequest {
  optional string tile_filetype = 1; // file type [e.g. tif, png, or jpg]
  optional int32 pixel_format   = 2; // see vw/Image/PixelTypeInfo.h
  optional int32 channel_type   = 3; // see vw/Image/PixelTypeInfo.h
  optional string type          = 4; // platefile type [ toast, kml, or gigapan ]
}

message IndexInfoRequest {
  required int32 platefile_id = 1;
}

message IndexReadRequest {
  required int32 platefile_id = 1;
  required int32 col = 3;
  required int32 row = 4;
  required int32 depth = 5;
  required int32 transaction_id = 6;
}

message IndexWriteRequest {
  required int32 platefile_id = 1;
  required int32 size = 3;
}

message IndexWriteComplete {
  required int32 platefile_id = 1;
  required TileHeader header = 3;
  required IndexRecord record = 4;
}

message IndexTransactionRequest {
  required int32 platefile_id = 1;
  required string description = 2;
  repeated TileHeader tile_headers = 3;
}

message IndexTransactionComplete {
  required int32 platefile_id = 1;
  required int32 transaction_id = 2;
}

message IndexTransactionCursorRequest {
  required int32 platefile_id = 1;
}

message IndexDepthRequest {
  required int32 platefile_id = 1;
}

// ----------------------------------
// Server --> Client Responses
// ----------------------------------

message RpcErrorMessage {
  required string type = 1;
  required string message = 2;
}

message IndexOpenReply {
  required string short_plate_filename = 1;
  required string full_plate_filename = 2;
  required IndexHeader index_header = 3;
}

message IndexInfoReply {
  required string short_plate_filename = 1;
  required string full_plate_filename = 2;
  required IndexHeader index_header = 3;
}

message IndexListReply {
  repeated int32 platefile_id = 1;
}

message IndexReadReply {
  required IndexRecord index_record = 1;
}

message IndexSuccess {
  optional string message = 1;
}

message IndexWriteReply {
  required int32 blob_id = 1;
}

message IndexTransactionReply {
  required int32 transaction_id = 1;
}

message IndexTransactionCursorReply {
  required int32 transaction_id = 1;
}

message IndexDepthReply {
  required int32 depth = 1;
}

// ---------------------------------
// Service definition
// ---------------------------------

message RpcErrorInfo {
  required string type = 1;
  required string message = 2;
}

message RpcNullMessage {
  optional int32 dummy = 1 [ default = 0];  // dummy parameter (empty message)
}

message RpcRequestWrapper {
  required string requestor = 1;
  required string method = 2;
  required bytes payload = 3;
}

message RpcResponseWrapper {
  optional bytes payload = 1;
  required bool error = 2 [ default = false ];
  optional RpcErrorInfo error_info = 3;
}

service IndexService {
  rpc OpenRequest (IndexOpenRequest) returns (IndexOpenReply);
  rpc CreateRequest (IndexCreateRequest) returns (IndexOpenReply);
  rpc InfoRequest (IndexInfoRequest) returns (IndexInfoReply);
  rpc ReadRequest (IndexReadRequest) returns (IndexReadReply);
  rpc WriteRequest (IndexWriteRequest) returns (IndexWriteReply);
  rpc WriteComplete (IndexWriteComplete) returns (RpcNullMessage);
  rpc TransactionRequest (IndexTransactionRequest) returns (IndexTransactionReply);
  rpc TransactionComplete (IndexTransactionComplete) returns (RpcNullMessage);
  rpc TransactionCursor (IndexTransactionCursorRequest) returns (IndexTransactionCursorReply);
  rpc DepthRequest (IndexDepthRequest) returns (IndexDepthReply);
  rpc ListRequest (IndexListRequest) returns (IndexListReply);
}

service IndexManagerService {
  rpc ListRequest (IndexListRequest) returns (IndexListReply);
}
